### PR TITLE
feat: Enable API Logging for Better Debugging POC

### PR DESCRIPTION
**Pull request by bkryu**
## 📌 Description
tl; dr: Current PR adds a logging system for input/output tracking to aid debugging FlashInfer APIs via a `@flashinfer_api` decorator.

**This PR does not label `@flashinfer_api` to every FlashInfer API -- many operations are missing labels. Further labeling is left for subsequent work.**

This PR introduces a production-ready API logging infrastructure that tracks function calls, arguments, and return values via a simple one-line decorator. Any function can be decorated with the decorator to track the input/output values in the API logger.

Key Features:
* Logging level controlled by `FLASHINFER_LOGLEVEL`
* Log destination set by `FLASHINFER_LOGDEST`; defaults to `stdout`
* Zero overhead when disabled (level 0 returns original function) as seen from `benchmarks/bench_logging_overhead.py`

Example usage
```
export FLASHINFER_LOGLEVEL=1
export FLASHINFER_LOGDEST="./flashinfer_api.log"

python3 benchmarks/flashinfer_benchmark.py --routine BatchDecodeWithPagedKVCacheWrapper --backends fa2 fa2_tc cudnn trtllm-gen trtllm-gen-native --page_size 16 --batch_size 1 --s_qo 1 --s_kv 1024 --num_qo_heads 64 --num_kv_heads 8 --head_dim_qk 128 --head_dim_vo 128 --random_actual_seq_len -vv --refcheck --q_dtype bfloat16 --kv_dtype bfloat16
```
produces log
```
================================================================================
[2025-11-20 17:51:18] FlashInfer API Logging - System Information
================================================================================
FlashInfer version: 0.5.2
CUDA toolkit version: 13.0
cuDNN version: 91600
Number of GPUs: 1
  GPU 0: NVIDIA B200
    Compute capability: 10.0 (SM100)
PyTorch version: 2.9.0+cu130
================================================================================

[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.__init__
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.plan
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.__init__
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.plan
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.__init__
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.plan
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.run
[2025-11-20 17:51:19] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.run
...

```

`export FLASHINFER_LOGLEVEL=3` produces:
```
(System Info same as above)
================================================================================
[2025-11-20 17:51:58] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.__init__
--------------------------------------------------------------------------------
Positional input arguments:
  arg[0]:
    <flashinfer.decode.BatchDecodeWithPagedKVCacheWrapper object at 0x1234399e3410>
  arg[1]:
    Tensor(
      shape=(134217728,)
      stride=(1,)
      dtype=torch.int8
      device=cuda:0
      requires_grad=False
      is_contiguous=True
    )
  arg[2]:
    'HND'
Keyword input arguments:
  use_cuda_graph=
    True
  use_tensor_cores=
    False
  paged_kv_indptr_buffer=
    Tensor(
      shape=(2,)
      stride=(1,)
      dtype=torch.int32
      device=cuda:0
      requires_grad=False
      is_contiguous=True
    )
  paged_kv_indices_buffer=
    Tensor(
      shape=(6,)
      stride=(1,)
      dtype=torch.int32
      device=cuda:0
      requires_grad=False
      is_contiguous=True
    )
  paged_kv_last_page_len_buffer=
    Tensor(
      shape=(1,)
      stride=(1,)
      dtype=torch.int32
      device=cuda:0
      requires_grad=False
      is_contiguous=True
    )
  backend=
    'fa2'
Default parameters (not explicitly provided):
  jit_args= [DEFAULT]
    None
Output value:
  None
================================================================================
...
```
`export FLASHINFER_LOGLEVEL=5` produces:
```
(System Info same as above)
================================================================================
[2025-11-20 17:52:23] FlashInfer API Call: BatchDecodeWithPagedKVCacheWrapper.__init__
--------------------------------------------------------------------------------
Positional input arguments:
  arg[0]:
    <flashinfer.decode.BatchDecodeWithPagedKVCacheWrapper object at 0x7a9fd9a88c0>
  arg[1]:
    Tensor(
      shape=(134217728,)
      stride=(1,)
      dtype=torch.int8
      device=cuda:0
      requires_grad=False
      is_contiguous=True
      min=0
      max=0
      mean=0.000000
    )
  arg[2]:
    'HND'
Keyword input arguments:
  use_cuda_graph=
    True
  use_tensor_cores=
    False
  paged_kv_indptr_buffer=
    Tensor(
      shape=(2,)
      stride=(1,)
      dtype=torch.int32
      device=cuda:0
      requires_grad=False
      is_contiguous=True
      min=0
      max=6
      mean=3.000000
    )
  paged_kv_indices_buffer=
    Tensor(
      shape=(6,)
      stride=(1,)
      dtype=torch.int32
      device=cuda:0
      requires_grad=False
      is_contiguous=True
      min=0
      max=5
      mean=2.500000
    )
  paged_kv_last_page_len_buffer=
    Tensor(
      shape=(1,)
      stride=(1,)
      dtype=torch.int32
      device=cuda:0
      requires_grad=False
      is_contiguous=True
      min=4
      max=4
      mean=4.000000
    )
  backend=
    'fa2'
Default parameters (not explicitly provided):
  jit_args= [DEFAULT]
    None
Output value:
  None
================================================================================
...
```

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Release Notes

* **New Features**
  * Added API logging feature configurable via environment variables (FLASHINFER_LOGLEVEL for level control, FLASHINFER_LOGDEST for destination)
  * Supports five verbosity levels with function names, inputs, outputs, metadata, and tensor statistics
  * Zero-overhead operation when disabled

* **Tests**
  * Added comprehensive logging test suite

* **Documentation**
  * Added logging configuration and usage documentation

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by bkryu** [comment]
> This is a great feature to have, and thanks @bkryu for brining this up! One general question, do you think we should log kernel behavior at python side or C++ side?

Thanks @yzh119. This is a question I considered, more from a "should we log what happens inside each function call because the current design only logs inputs and outputs" point of view. My initial sentiment is that logging kernel behavior will be tricky because we have multiple kernel providers -- some of which are closed source and others are OSS, and some that already come with logging:
* [cuDNN](https://docs.nvidia.com/deeplearning/cudnn/backend/latest/reference/troubleshooting.html#error-reporting-and-api-logging) and [cuBLAS](https://docs.nvidia.com/cuda/cublas/#cublasloggerconfigure) has their own respective API logging settings for logging kernel and kernel selecting heuristic behavior. For these backends, it is best if we use their API logs to debug them.
* [TRT-LLM](https://nvidia.github.io/TensorRT-LLM/reference/troubleshooting.html#debug-execution-errors) appears to have something but I am not sure whether it has the ability to log kernel-level info useful for debugging.
* FlashInfer-native kernels (CUDA and CuTe DSL) do not have a logging infrastructure and will require us to implement something.
That being said, if we believe that logging kernel behavior is useful, I agree that we can look into it

**Comment by yzh119** [comment]
@xslingcn @Ubospica @yiyanzhai @zanderjiang this PR should greatly reducing the overhead of extracting workloads for flashinfer-bench project.

**Comment by Ubospica** [comment]
@yzh119 Thanks for sharing! The low-overhead design looks very useful. We definitely have plans to support low-overhead dump and CUDA graph support in our roadmap. I will leverage this PR as a key reference.

File: README.md [hunk 1]
```diff
@@ -169,6 +169,20 @@ o = flashinfer.single_prefill_with_kv_cache(q, k, v, causal=False) # prefill att
 
 Check out [documentation](https://docs.flashinfer.ai/) for usage of batch decode/append/prefill kernels and shared-prefix cascading kernels.
 
+## API Logging
+
+FlashInfer provides comprehensive API logging for debugging. Enable it using environment variables:
+
+```bash
+# Enable logging (levels: 0=off (default), 1=basic, 3=detailed, 5=statistics)
+export FLASHINFER_LOGLEVEL_DBG=3
```
Comment (aleozlx) [inline README.md]: why do we always have "_DBG"?

File: LOGGING.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# FlashInfer Logging
```
Comment (aleozlx) [inline LOGGING.md]: do we wanna organize this file in docs/ etc?

cc @yzh119

File: README.md [hunk 3]
```diff
@@ -169,6 +169,20 @@ o = flashinfer.single_prefill_with_kv_cache(q, k, v, causal=False) # prefill att
...
+```bash
+# Enable logging (levels: 0=off (default), 1=basic, 3=detailed, 5=statistics)
+export FLASHINFER_LOGLEVEL=3
+
+# Set log destination (stdout (default), stderr, or file path)
+export FLASHINFER_LOGDEST=stdout
+```
+
+For detailed information about logging levels, configuration, and advanced features, see [LOGGING.md](LOGGING.md).
+
```
[the remark this answers is not in the document]
Comment (bkryu) [reply README.md:185]: Was a typo. fixed in latest commit

File: README.md (hunk 1 above)
Comment (bkryu) [reply README.md]: Removed the _DBG suffix

File: LOGGING.md (hunk 2 above)
Comment (bkryu) [reply LOGGING.md]: Or might be even better to convert this to an .rst so that it appears in the [documentations](https://docs.flashinfer.ai/) 

@yzh119 do you have suggestions on how I might convert and where I can place the file?

**Comment by bkryu** [comment]
> @xslingcn @Ubospica @YiyanZhai @zanderjiang this PR should greatly reducing the overhead of extracting workloads for flashinfer-bench project.

Hi @xslingcn @Ubospica @YiyanZhai @zanderjiang, it would be useful if there are any suggestions on the logging behavior that can benefit flashinfer-bench. I am leaving a logging level for dumping tensors for future work, but if there are any please feel free to provide any input.

File: flashinfer/api_logging.py [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+    # Create handler based on destination
+    if _API_LOG_DEST == "stdout":
+        handler = logging.StreamHandler(sys.stdout)
+    elif _API_LOG_DEST == "stderr":
+        handler = logging.StreamHandler(sys.stderr)
+    else:
+        handler = logging.FileHandler(_API_LOG_DEST, mode="a")
+
+    # Use a simple formatter (we'll add timestamps manually to key lines)
+    formatter = logging.Formatter("%(message)s")
```
Comment (aleozlx) [inline flashinfer/api_logging.py:72]: should we expose the formatting as an env var to allow pid time stamps etc overridable by the user?

File: flashinfer/api_logging.py [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+
+# Helper function to substitute %i with process ID in file paths
+def _substitute_process_id(path: str) -> str:
+    """
+    Replace %i with the current process ID in a path.
+
+    This is useful for multi-process/multi-GPU environments where each process
+    needs its own log file.
+    """
+    if "%i" in path:
```
Comment (aleozlx) [inline flashinfer/api_logging.py:36]: i think the standard library has 

%(process)d
https://docs.python.org/3/library/logging.html#logrecord-attributes

we don't have to reinvent new things contrary to established standards

**Comment by aleozlx** [comment]
(non-blocking) we can keep `@flashinfer_log` for cases where we do wanna log helper functions
cc @bkryu

File: flashinfer/api_logging.py (hunk 5 above)
Comment (aleozlx) [reply flashinfer/api_logging.py:36]: oh this is about file name, i was probably mistaken...

Comment (bkryu) [reply flashinfer/api_logging.py:36]: Yes I think the %(process)d only working in the logging module at the moment

File: LOGGING.md (hunk 2 above)
Comment (yzh119) [reply LOGGING.md]: I think placing it under docs/logging.rst and indexing it in docs/index.rst should be great.

>  how I might convert

I have tried pandoc before but I believe any existing llm (gemini/claude/gpt) could do a better job.

**Review by yzh119 — APPROVED** [review]
LGTM overall, thanks @bkryu for the great work!

File: LOGGING.md (hunk 2 above)
Comment (bkryu) [reply LOGGING.md]: @yzh119, makes sense. I converted in the latest commit and built the sphinx documentation locally

Here is how the index page looks: See red marks for where I placed it:
Screenshot 2025-11-21 at 5 17 47 PM

and here is how the actual logging page looks:
Screenshot 2025-11-21 at 5 18 19 PM
